org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/generator/trace/ITraceForStorageProviderExtension.java

```diff
@@ -0,0 +1,36 @@
+/*******************************************************************************
+ * Copyright (c) 2017 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.ui.generator.trace;
+
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IStorage;
+
+/**
+ * @author Karsten Thoms (karsten.thoms@itemis.de) - Initial contribution and API
+ * @since 2.13
+ */
+public interface ITraceForStorageProviderExtension {
+	/**
+	 * Returns the trace file for the given derived resource. May return
+	 * <code>null</code> if no such file is available for the given storage.
+	 * 
+	 * @param generatedFile
+	 *            the resource whose trace is requested. May not be <code>null</code>.
+	 * @return the trace file for <code>generatedFile</code> or <code>null</code>.
+	 */
+	/*@Nullable*/
+	IFile getTraceFile(/*@NonNull*/ IFile generatedFile);
```

Comment:
Could these become default methods on the ITraceForStorageProvider? This would greatly simplify the implementation.